clients/python-wrapper/lakefs/branch.py

```diff
@@ -170,79 +236,128 @@ def import_data(self, commit_message: str = "", metadata: Optional[dict] = None)
+        3. Commit changes
+        4. Merge back to source branch
+        5. Delete ephemeral branch
 
-        This can also be used in combination with object listing. For example:
+        Using a transaction the code for this flow will look like this:
 
         .. code-block:: python
 
             import lakefs
 
             branch = lakefs.repository("<repository_name>").branch("<branch_name>")
-            # list objects on a common prefix
-            objs = branch.objects(prefix="my-object-prefix/", max_amount=100)
-            # delete objects which have "foo" in their name
-            branch.delete_objects([o.path for o in objs if "foo" in o.path])
+            with branch.transact(commit_message="my transaction") as tx:
+                for obj in tx.objects(prefix="prefix_to_delete/"):  # Delete some objects
+                    obj.delete()
 
-        :param object_paths: a single path or an iterable of paths to delete
-        :raise NotFoundException: if branch or repository do not exist
-        :raise NotAuthorizedException: if user is not authorized to perform this operation
-        :raise ServerException: for any other errors
+                # Create new object
+                tx.object("new_object").upload("new object data")
+
+        Note that unlike database transactions, lakeFS transaction does not take a "lock" on the branch, and therefore
+        the transaction might fail due to changes in source branch after the transaction was created.
+
+        :param commit_message: once the transaction is committed, a commit is created with this message
+        :param commit_metadata: user metadata for the transaction commit
+        :return: a Transaction object to perform the operations on
         """
-        if isinstance(object_paths, str):
-            object_paths = [object_paths]
-        elif isinstance(object_paths, StoredObject):
-            object_paths = [object_paths.path]
-        elif isinstance(object_paths, Iterable):
-            object_paths = [o.path if isinstance(o, StoredObject) else o for o in object_paths]
-        with api_exception_handler():
-            return self._client.sdk_client.objects_api.delete_objects(
-                self._repo_id,
-                self._id,
-                lakefs_sdk.PathList(paths=object_paths)
-            )
+        with Transaction(self._repo_id, self._id, commit_message, commit_metadata, self._client) as tx:
+            yield tx
+
+
+class _Transaction(_BaseBranch):
+    @staticmethod
+    def _get_tx_name() -> str:
+        return f"tx-{uuid.uuid4()}"  # Don't rely on source branch name as this might exceed valid branch length
+
+    def __init__(self, repository_id: str, branch_id: str, commit_message: str = "",
+                 commit_metadata: Optional[Dict] = None, client: Client = None):
+        self._commit_message = commit_message
+        self._commit_metadata = commit_metadata
+        self._source_branch = branch_id
```

Comment:
Is it allowed to create a transaction on branches with uncommitted changes? if so, it might be a bad experience noticing only when trying to complete the transaction and failing to merge into a branch with uncommitted changes.

Reply:
Good feedback. On the other hand, dirty branch can be committed before transaction is completed, and transaction created on not dirty branch does not guarantee branch will not be dirty when trying to commit the transaction.
@ozkatz You thoughts?